### PR TITLE
HDDS-7630. Show detailed OMResponse when OzoneManagerDoubleBuffer terminates

### PR DESCRIPTION
**Pull request by symious**
## What changes were proposed in this pull request?

When OzoneManagerDoubleBuffer fails due to exception, it's better to have detailed information of the OMResponse for debugging.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7630

## How was this patch tested?

no test.

Some example outputs:

* Create bucket
```
cmdType: CreateBucket
traceID: ""
success: true
status: OK
createBucketResponse {
}
```
* Create key
```
cmdType: CreateKey
traceID: ""
success: true
status: OK
createKeyResponse {
  keyInfo {
    volumeName: "vol1"
    bucketName: "bucket1"
    keyName: "key1"
    dataSize: 3879
    type: RATIS
    factor: THREE
    keyLocationList {
      version: 0
      keyLocations {
        blockID {
          containerBlockID {
            containerID: 1
            localID: 109611004723200001
          }
          blockCommitSequenceId: 0
        }
        offset: 0
        length: 268435456
        createVersion: 0
        pipeline {
          members {
            uuid: "c6d47952-d622-4723-b28b-a6ff48fc1fe0"
            ipAddress: "172.21.0.8"
            hostName: "ozone-ha_datanode_2.ozone-ha_default"
            ports {
              name: "REPLICATION"
              value: 9886
            }
            ports {
              name: "RATIS"
              value: 9858
            }
            ports {
              name: "RATIS_ADMIN"
              value: 9857
            }
            ports {
              name: "RATIS_SERVER"
              value: 9856
            }
            ports {
              name: "STANDALONE"
              value: 9859
            }
            networkName: "c6d47952-d622-4723-b28b-a6ff48fc1fe0"
            networkLocation: "/default-rack"
            persistedOpState: IN_SERVICE
            persistedOpStateExpiry: 0
            uuid128 {
              mostSigBits: -4119534362450311389
              leastSigBits: -5581183697824243744
            }
          }
          members {
            uuid: "5a352379-c80a-4baa-aac7-fba7ea627a93"
            ipAddress: "172.21.0.4"
            hostName: "ozone-ha_datanode_4.ozone-ha_default"
            ports {
              name: "REPLICATION"
              value: 9886
            }
            ports {
              name: "RATIS"
              value: 9858
            }
            ports {
              name: "RATIS_ADMIN"
              value: 9857
            }
            ports {
              name: "RATIS_SERVER"
              value: 9856
            }
            ports {
              name: "STANDALONE"
              value: 9859
            }
            networkName: "5a352379-c80a-4baa-aac7-fba7ea627a93"
            networkLocation: "/default-rack"
            persistedOpState: IN_SERVICE
            persistedOpStateExpiry: 0
            uuid128 {
              mostSigBits: 6500140643133311914
              leastSigBits: -6140662868285949293
            }
          }
          members {
            uuid: "bc5e9247-5425-4f3a-a46f-22f449a0c94e"
            ipAddress: "172.21.0.3"
            hostName: "ozone-ha_datanode_3.ozone-ha_default"
            ports {
              name: "REPLICATION"
              value: 9886
            }
            ports {
              name: "RATIS"
              value: 9858
            }
            ports {
              name: "RATIS_ADMIN"
              value: 9857
            }
            ports {
              name: "RATIS_SERVER"
              value: 9856
            }
            ports {
              name: "STANDALONE"
              value: 9859
            }
            networkName: "bc5e9247-5425-4f3a-a46f-22f449a0c94e"
            networkLocation: "/default-rack"
            persistedOpState: IN_SERVICE
            persistedOpStateExpiry: 0
            uuid128 {
              mostSigBits: -4873296911716233414
              leastSigBits: -6598016496471848626
            }
          }
          state: PIPELINE_OPEN
          type: RATIS
          factor: THREE
          id {
            id: "38cafb50-39f9-407c-948b-fb8c8dfa1c0a"
            uuid128 {
              mostSigBits: 4092359533408108668
              leastSigBits: -7742818553237464054
            }
          }
          leaderID: "bc5e9247-5425-4f3a-a46f-22f449a0c94e"
          creationTimeStamp: 1671504330972
          suggestedLeaderID {
            mostSigBits: -4873296911716233414
            leastSigBits: -6598016496471848626
          }
          memberReplicaIndexes: 0
          memberReplicaIndexes: 0
          memberReplicaIndexes: 0
          leaderID128 {
            mostSigBits: -4873296911716233414
            leastSigBits: -6598016496471848626
          }
        }
        partNumber: 0
      }
      isMultipartKey: false
    }
    creationTime: 1671504595449
    modificationTime: 1671504595449
    latestVersion: 0
    acls {
      type: USER
      name: "hadoop"
      rights: "\200"
      aclScope: ACCESS
    }
    acls {
      type: GROUP
      name: "hadoop"
      rights: "\200"
      aclScope: ACCESS
    }
    objectID: 9223372036854778112
    updateID: 9
    parentID: 0
    isFile: false
  }
  ID: 109543725169115136
  openVersion: 0
}

```

### CODE REVIEW

**Comment by kerneltime** [comment]
@symious can a user cause OM logs to get flooded with response errors?

**Comment by kerneltime** [comment]
cc @errose28 @duongkame can you please take a look.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java [hunk 1]
```diff
@@ -264,6 +264,7 @@ private Void addToBatchTransactionInfoWithTrace(String parentName,
    */
   private void flushTransactions() {
     while (isRunning.get()) {
+      AtomicReference<OMResponse> omResponse = new AtomicReference<>();
```
Comment (errose28) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:267]: Why is this needed?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java [hunk 2]
```diff
@@ -480,9 +482,10 @@ public void stop() {
 
   }
 
-  private void terminate(IOException ex) {
+  private void terminate(IOException ex, OMResponse omResponse) {
     String message = "During flush to DB encountered error in " +
-        "OMDoubleBuffer flush thread " + Thread.currentThread().getName();
+        "OMDoubleBuffer flush thread " + Thread.currentThread().getName() +
+        " when handling OMRequest: " + omResponse;
```
Comment (errose28) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:488]: What string does this print for the OMResponse? I'm not sure it will actually be the name of the response proto message.

**Comment by errose28** [comment]
@kerneltime users should not be able to flood the logs with this change. It only modifies existing log messages instead of adding new ones. The messages that are modified happen when there is a fatal error flushing the double buffer. This is after any error caused by a user would be processed in the OMRequest validation.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java (hunk 2 above)
Comment (symious) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:488]: Updated in the description.
I got the output by just print the Response before handling it.

**Review by errose28 — COMMENTED** [review]
The OM response is only in scope as part of the lambda from line 277 to 295. The try/catch block there should handle any exceptions specifically related to that response. The outer try/catch would catch more general failures not related to the response, so response logging should not be necessary in that part. I think an approach like [this](https://github.com/errose28/ozone/commit/c0fe3096e0581f1e8dfec3a7d52193a6617a04cf) might be better.

**Comment by symious** [comment]
@errose28 In our cluster, we had the following exception.
```
2022-12-21 09:20:06,060 [OMDoubleBufferFlushThread] ERROR org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer: Terminating with exit status 2: OMDoubleBuffer flush threadOM
DoubleBufferFlushThreadencountered Throwable error
java.lang.NullPointerException
        at org.apache.hadoop.ozone.om.helpers.RepeatedOmKeyInfo.getProto(RepeatedOmKeyInfo.java:76)
        at org.apache.hadoop.ozone.om.codec.RepeatedOmKeyInfoCodec.toPersistedFormat(RepeatedOmKeyInfoCodec.java:50)
```

Since it's not IOException, it will be catched in the outer try-catch, maybe we should also keep the change for outer "throwable" catch?

**Comment by errose28** [comment]
Once the outer try/catch is hit, the specific response is out of scope. If we wanted response information for the example above, we could add a separate `Throwable` catch to the inner try/catch block and print the response there, similar to what is done in the outer try/catch.

**Comment by symious** [comment]
```
2022-07-22 15:20:39,215 [OMDoubleBufferFlushThread] ERROR org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer: Terminating with exit status 2: OMDoubleBuffer flush threadOMDoubleBufferFlushThreadencountered Throwable error
java.util.ConcurrentModificationException
        at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
        at java.util.ArrayList$Itr.next(ArrayList.java:859)
        at org.apache.hadoop.ozone.om.helpers.RepeatedOmKeyInfo.getProto(RepeatedOmKeyInfo.java:70)
        at org.apache.hadoop.ozone.om.codec.RepeatedOmKeyInfoCodec.toPersistedFormat(RepeatedOmKeyInfoCodec.java:50)
        at org.apache.hadoop.ozone.om.codec.RepeatedOmKeyInfoCodec.toPersistedFormat(RepeatedOmKeyInfoCodec.java:35)
        at org.apache.hadoop.hdds.utils.db.CodecRegistry.asRawData(CodecRegistry.java:85)
        at org.apache.hadoop.hdds.utils.db.TypedTable.putWithBatch(TypedTable.java:130)
        at org.apache.hadoop.ozone.om.response.key.OMKeyCommitResponse.updateDeletedTable(OMKeyCommitResponse.java:107)
        at org.apache.hadoop.ozone.om.response.key.OMKeyCommitResponse.addToDBBatch(OMKeyCommitResponse.java:80)
        at org.apache.hadoop.ozone.om.response.OMClientResponse.checkAndUpdateDB(OMClientResponse.java:68)
        at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.lambda$2(OzoneManagerDoubleBuffer.java:256)
        at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.addToBatchWithTrace(OzoneManagerDoubleBuffer.java:201)
        at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.lambda$1(OzoneManagerDoubleBuffer.java:254)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.flushTransactions(OzoneManagerDoubleBuffer.java:250)
        at java.lang.Thread.run(Thread.java:748)
```
@errose28 , I added the whole stacktrace, I think the exception is thrown in the for loop, so the omResponse can still hold the information?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java [hunk 3]
```diff
@@ -276,9 +277,9 @@ private void flushTransactions() {
             AtomicReference<String> lastTraceId = new AtomicReference<>();
             readyBuffer.iterator().forEachRemaining((entry) -> {
               try {
-                OMResponse omResponse = entry.getResponse().getOMResponse();
```
Comment (duongkame) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:279]: nit: I don't think `omResponse` is used in any lambda. You probably just need to move this line to outside the try block (move up 2 lines). Another `AtomicReference ` is not needed.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java [hunk 4]
```diff
@@ -386,11 +389,14 @@ private void flushTransactions() {
               + "exit.", Thread.currentThread().getName());
         }
       } catch (IOException ex) {
-        terminate(ex);
+        terminate(ex, omResponse.get());
       } catch (Throwable t) {
-        final String s = "OMDoubleBuffer flush thread " +
-            Thread.currentThread().getName() + " encountered Throwable error";
-        ExitUtils.terminate(2, s, t, LOG);
+        StringBuilder s = new StringBuilder("OMDoubleBuffer flush thread " +
```
Comment (duongkame) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:394]: Guess if `OzoneManagerDoubleBuffer#terminate` took exitCode as a parameter, you'd just need to replace the entire thing here with `terminate(ex, omResponse.get(), 2); `

**Review by duongkame — APPROVED** [review]
I'm on board with the idea of logging more details for a termination. +1.
I left a couple of minor comments you can address optionally.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java (hunk 4 above)
Comment (symious) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java]: @duongkame Thank you for the suggestions, updated the patch.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java (hunk 3 above)
Comment (symious) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:279]: I'm not sure if I'm understanding correctly, this is the inital error IDE gives me.
Screenshot 2023-01-05 at 10 33 18 AM

Comment (duongkame) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java:279]: I was wrong actually, thought it `omResponse` is used only inside the for loop. It's alright then.

**Comment by errose28** [comment]
> I added the whole stacktrace, I think the exception is thrown in the for loop, so the omResponse can still hold the information?

Right, the exception is thrown in the loop, but it is not caught until the outer try/catch, where the response has gone out of scope in the original code. Instead of changing the scope of the response, we can add a more generic catch inside the loop as well. The try/catch inside the loop would handle exceptions related to applying that specific response. The outer try/catch would handle general state machine exceptions, like committing the whole batch.

**Comment by symious** [comment]
@errose28 Got the idea, update the patch, could you have a look?

**Review by errose28 — COMMENTED** [review]
Thanks for the update @symious. The approach looks good now, I just have a minor suggestion for an improvement [here](https://github.com/errose28/ozone/commit/22fa9ccb8f22858faf7a1dbe039ab9d01f1372d3) which does not need the `AtomicReference`.

**Comment by symious** [comment]
@errose28 Thank you for the advice and patch updated, the inner catch of Throwable does make things easier.

**Review by kerneltime — APPROVED** [review]

**Review by errose28 — APPROVED** [review]

**Comment by errose28** [comment]
Thanks for the patch @symious
